### PR TITLE
NaiveBayesTrainer and NaiveBayesClassifier

### PR DESCRIPTION
**Pull request by anwar6953**
NaiveBayesTrainer will train based upon a given String field and prediction label in a record and save this model to a FileSet.

NaiveBayesClassifier will load this model and tag records with a prediction label, based upon a String in a particular field.

Depends on:
https://github.com/caskdata/cdap/pull/5407

Pending: Docs, Widgets.

### CODE REVIEW

File: spark-plugins/pom.xml [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.felix</groupId>
+        <artifactId>maven-bundle-plugin</artifactId>
+        <version>2.5.4</version>
+        <extensions>true</extensions>
+        <configuration>
+          <instructions>
+            <_exportcontents>co.cask.hydrator.plugin.batch.spark.*;org.apache.commons.lang;org.apache.hive.hcatalog.*
```
Comment (albertshau) [inline spark-plugins/pom.xml]: should be able to remove commons.lang and hcatalog

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+   */
+  public static class Config extends PluginConfig {
+
+    @Description("FileSet to use to load the model from.")
+    private final String fileSetName;
+
+    @Description("Path of the FileSet to load the model from.")
+    private final String path;
+
+    @Description("A space-separated sequence of words, which to classify.")
```
Comment (albertshau) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java]: nit: 'sequence of words to classify'

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    private final String fieldToClassify;
+
+    @Description("The field on which to set the prediction. It must be of type double.")
```
Comment (albertshau) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java]: nit: 'It will be of type double"

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+          .set(config.fieldToSet, prediction)
+          .build();
+      }
+    });
+    return output;
+  }
+
+  // creates a builder based off the given record
+  private StructuredRecord.Builder cloneRecord(StructuredRecord record) {
+    Schema schemaToUse = outputSchema != null ? outputSchema : record.getSchema();
```
Comment (anwar6953) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java]: err, this is wrong. I need to not directly use the record.getSchema, but add a field to it first.

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesTrainer.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+   */
+  public static class Config extends PluginConfig {
+
+    @Description("FileSet to use to load the model from.")
+    private final String fileSetName;
+
+    @Description("Path of the FileSet to load the model from.")
+    private final String path;
+
+    @Description("A space-separated sequence of words, which to use for classification.")
```
Comment (albertshau) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesTrainer.java]: nit: 'sequence of words to use for'

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesTrainer.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    }
+  }
+
+  public NaiveBayesTrainer(Config config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    pipelineConfigurer.createDataset(config.fileSetName, FileSet.class, FileSetProperties.builder()
```
Comment (albertshau) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesTrainer.java]: should do some validation too.  For example, if the input schema is non-null, make sure the fieldToClassify exists and is of type String, and the predictionField exists and is of type double

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/SpamMessage.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a string and whether that is spam or not.
+ */
+public class SpamMessage {
```
Comment (albertshau) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/SpamMessage.java]: this is purely for unit tests right? If so we can move it to the test directory rather than the main directory.

File: spark-plugins/src/test/java/co/cask/hydrator/plugin/batch/spark/testclasses/MockSource.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    String recordStr = input.getValue().getString(RECORD_COL);
+    emitter.emit(StructuredRecordStringConverter.fromJsonString(recordStr, schema));
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    context.setInput(config.tableName);
+  }
+
+  public static co.cask.cdap.etl.proto.v2.ETLPlugin getPlugin(String tableName) {
```
Comment (albertshau) [inline spark-plugins/src/test/java/co/cask/hydrator/plugin/batch/spark/testclasses/MockSource.java]: don't need the fully qualified name (this is why we renamed it from v1.Plugin to v2.ETLPlugin)

File: spark-plugins/src/test/java/co/cask/hydrator/plugin/batch/spark/testclasses/MockSink.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<byte[], Put>> emitter) throws Exception {
+    byte[] rowkey = Bytes.toBytes(UUID.randomUUID());
+    Put put = new Put(rowkey);
+    put.add(SCHEMA_COL, input.getSchema().toString());
+    put.add(RECORD_COL, StructuredRecordStringConverter.toJsonString(input));
+    emitter.emit(new KeyValue<>(rowkey, put));
+  }
+
+  public static co.cask.cdap.etl.proto.v2.ETLPlugin getPlugin(String tableName) {
```
Comment (albertshau) [inline spark-plugins/src/test/java/co/cask/hydrator/plugin/batch/spark/testclasses/MockSink.java]: dont need fully qualified name

**Comment by albertshau** [comment]
a few comments, basic implementation looks good

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesTrainer.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.mllib.classification.NaiveBayes;
+import org.apache.spark.mllib.classification.NaiveBayesModel;
+import org.apache.spark.mllib.feature.HashingTF;
+import org.apache.spark.mllib.regression.LabeledPoint;
+
+/**
+ * Spark Sink plugin that trains a model based upon whether messages are spam or not, and then classifies messages.
```
Comment (gokulavasan) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesTrainer.java]: Is this Javadoc right? This seems to be the Javadoc for classifier and not the trainer?

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesTrainer.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+  public static final String PLUGIN_NAME = "NaiveBayesTrainer";
+
+  private final Config config;
+
+  /**
+   * Configuration for the NaiveBayesTrainer.
+   */
+  public static class Config extends PluginConfig {
+
+    @Description("FileSet to use to load the model from.")
```
Comment (gokulavasan) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesTrainer.java]: I thought we save the model to the FileSet?

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java [hunk 12]
```diff
@@ -83,11 +83,6 @@ public Config(String fileSetName, String path, String fieldToClassify, String fi
     }
   }
 
-  // for unit tests, otherwise config is injected by plugin framework.
-  public NaiveBayesClassifier(Config config) {
```
Comment (albertshau) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java]: for future reference, its ok to have this too, plugin framework calls this constructor if it exists.

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java [hunk 13]
```diff
@@ -99,14 +94,25 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws Ille
...
   }
 
+  private void validateSchema(Schema inputSchema) {
+    Schema.Type fieldToClassifyType = inputSchema.getField(config.fieldToClassify).getSchema().getType();
+    Preconditions.checkArgument(fieldToClassifyType == Schema.Type.STRING,
+                                "Field to classify must be of type String, but was %s.", fieldToClassifyType);
+
+    Schema.Field predictionField = inputSchema.getField(config.fieldToSet);
+    Preconditions.checkArgument(predictionField == null,
+                                "Prediction field must already not exist in input schema.");
```
Comment (albertshau) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java]: nit: 'must not already'

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java (hunk 12 above)
Comment (anwar6953) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/NaiveBayesClassifier.java]: yea, I just thought it was unnecessary (since we dont use it in test cases), so just removed it.

**Comment by albertshau** [comment]
:+1:

**Comment by anwar6953** [comment]
Squashed commits; merging
